libhl-crypto/src/signatures/secp256k1.rs

```diff
@@ -0,0 +1,528 @@
+        bin2hex(&self.value[..])
+    }
+
+    pub fn from_slice(data: &[u8]) -> Result<EcdsaSecp256K1Sha256PrivateKey, CryptoError> {
+        if data.len() != PRIVATE_KEY_SIZE {
+            Err(CryptoError::KeyGenError(format!("Expected {} bytes for a private key", PRIVATE_KEY_SIZE)))
+        } else {
+            let mut value = [0u8; PRIVATE_KEY_SIZE];
+            array_copy!(data, value);
+            Ok(EcdsaSecp256K1Sha256PrivateKey { value })
+        }
+    }
+
+    pub fn from_hex(data: &str) -> Result<EcdsaSecp256K1Sha256PrivateKey, CryptoError> {
+        let bin = hex2bin(data)?;
+        EcdsaSecp256K1Sha256PrivateKey::from_slice(bin.as_slice())
+    }
+}
+
+impl Clone for EcdsaSecp256K1Sha256PrivateKey {
+    fn clone(&self) -> EcdsaSecp256K1Sha256PrivateKey {
+        EcdsaSecp256K1Sha256PrivateKey::from_slice(&self.value[..]).unwrap()
+    }
+}
+
+impl fmt::Display for EcdsaSecp256K1Sha256PrivateKey {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "EcdsaSecp256K1Sha256PrivateKey {{ value: {} }}", self.as_hex())
+    }
+}
+
+impl fmt::Debug for EcdsaSecp256K1Sha256PrivateKey {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "EcdsaSecp256K1Sha256PrivateKey {{ value: {} }}", self.as_hex())
+    }
+}
+
+impl Eq for EcdsaSecp256K1Sha256PrivateKey {}
+
+impl PartialEq for EcdsaSecp256K1Sha256PrivateKey {
+    fn eq(&self, other: &EcdsaSecp256K1Sha256PrivateKey) -> bool {
+        array_compare(&self.value, &other.value)
+    }
+}
+
+impl PrivateKey<EcdsaSecp256K1Sha256PublicKey> for EcdsaSecp256K1Sha256PrivateKey {
+    fn get_algorithm_name(&self) -> &str {
+        ALGORITHM_NAME
+    }
+
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, CryptoError> {
+        let mut r = [0u8; SIGNATURE_POINT_SIZE];
+        let mut s = [0u8; SIGNATURE_POINT_SIZE];
+        let mut pool = RAND::new();
+        let mut seed = [0u8; 128];
+        get_random_seed(&mut seed)?;
+        pool.seed(seed.len(), &seed);
+        ecdh::ecpsp_dsa(ecdh::SHA256, &mut pool, &self.value, &message, &mut r, &mut s);
+
+        //Use the "low s" form to be compatible with libsecp256k1
+        normalize_s(&mut s);
+        let mut signature = Vec::new();
+        signature.extend_from_slice(&r[..]);
+        signature.extend_from_slice(&s[..]);
+
+        Ok(signature)
+    }
+
+    fn get_public_key(&self) -> Result<EcdsaSecp256K1Sha256PublicKey, CryptoError> {
+        let mut w = [0u8; PUBLIC_KEY_SIZE]; //Compressed
+        let mut s = [0u8; PRIVATE_KEY_SIZE];
+        array_copy!(self.value, s);
+        ecdh::key_pair_generate(None, &mut s, &mut w);
+        zero!(s);
+        if ecdh::public_key_validate(&w) == 0 {
+            Ok(EcdsaSecp256K1Sha256PublicKey { value: w })
+        } else {
+            Err(CryptoError::KeyGenError("Invalid private key".to_string()))
+        }
+    }
+}
+
+fn normalize_s(s: &mut [u8; 32]) {
```

Comment:
A comment indicating the significance of low S form would be good. This might help https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-September/013098.html

Reply:
Also adding tests for `normalize_s` and `set_b32`, `is_high`, etc would be useful

Reply:
Done.